drivers/wifi/simplelink/simplelink_sockets.c

```diff
@@ -20,13 +20,24 @@ LOG_MODULE_DECLARE(LOG_MODULE_NAME);
 #include <ti/drivers/net/wifi/source/driver.h>
 #include "simplelink_support.h"
 
+#include "sockets_internal.h"
 #include "tls_internal.h"
 
 #define FAILED (-1)
 
+/* Socket LUT, matching system file descriptor with offloaded descriptor.
+ * Note: system FD will always be less than CONFIG_POSIX_MAX_FDS.
+ */
+static int simplelink_offload_fd[CONFIG_POSIX_MAX_FDS] = {
```

Comment:
So, fairly speaking, fdtable mini-subsys was designed to hold fd to the underlying object mapping and preclude the need for such duplicate mappings. Can you explain why it's needed? If https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/fdtable.h API is not flexible enough, we'd rather look into extending it, than add resource duplication like that. And if it's truly needed, then there rather be comment why it's needed and can't just use fdtable API.

Reply:
So the reason I need this mapping is that SimpleLink (and nRF91 as well) offloaded socket engines keep their own track of the socket descriptors assigned. These engines return "their" socket descriptor when socket is created, completely unaware from what's happening on the Zephyr side.

So I'm afraid it's not a matter of fdtable API flexibility.  Even if we could request a specific fd number from Zephyr, matching the one returned by the offloaded engine, we could end-up in a situation, that this specific fd is already in use on the  Zephyr side. That's why decided to have this LUT, to keep track of what Zephyr file descriptors are assigned to which offloaded socket descriptors.

Theoretically, I could hack the offloaded socket descriptor value into the `void *obj` field within the `struct fd_entry`, but again, I'd consider it a hack. As long as we treat `void *obj` as a pointer, we need to allocate the memory somewhere to keep the offloaded socket descriptor, that's what I use the LUT for as well.

Reply:
> So the reason I need this mapping is that SimpleLink (and nRF91 as well) offloaded socket engines keep their own track of the socket descriptors assigned. These engines return "their" socket descriptor when socket is created, completely unaware from what's happening on the Zephyr side.

Absolutely. This is no way different from native Zephyr networking API, which has "its" net_context*, and fdtable mini-subsys maps POSIX-level file descriptor to an arbitrary void* of the underlying implementation. So, as long as you need to data which is castable to void*, just use fdtable directly. If you need to store more data than that - then year, you'd need to define a struct for that data, and keep "shadow array" of such structs. My understanding, is that  "modem" offloading from this PR does that.

> Theoretically, I could hack the offloaded socket descriptor value into the void *obj field within the struct fd_entry, but again, I'd consider it a hack. 

It's not a hack, is the intended usage. void* is effectively a datum of size of bitness of the target platform. If it's not pure enough for you, feel free to submit a patch to replace it with explicit uintptr_t. But that way, we'll just have more casts around overall.

Reply:
Ok, if it's not against the original concept to reuse the `obj` field for such purpose (as storing some custom integer value), then I see no reason to insist on this extra array. I'll rework this as well.

Reply:
> if it's not against the original concept

Nope, it was exactly the opposite. And I appreciate this discussion, just the same as your work on plugging TLS into this system, which allowed to extend it and clarify some aspects. It's a matter of fact that there're still not clear enough things in there. Which reminds me that we had similar situation with @jukkar in other patch, and since then, it's in my TODO to actually add more comments to all this stuff to clarify it's "internal API" (that's part of the issue, IMHO these matters shouldn't be in user-level docs, why it wasn't commented thoroughly in the first place).

So, these discussions are important, please provide any suggestions on how to clarify this "internal API contract", feel free to improve smaller things, and bigger ones which would sidetrack you too much, feel free to emphasize to me, and I intend to get to improve it.

Thanks!